emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala

```diff
@@ -313,6 +318,34 @@ trait DataBag[A] extends Serializable {
           }
       })._1
   }
+
+  // -----------------------------------------------------
+  // equals, hashCode and toString
+  // -----------------------------------------------------
+
+  override def equals(o: Any) = o match {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
+        (thisVals.toSet diff thatVals.toSet).isEmpty
+    }
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    scala.util.hashing.MurmurHash3.unorderedHash(fetch())
```

Comment:
Maybe do it with a `fold`, see https://gist.github.com/aalexandrov/089f0fd4d3de13d37536f035dfe14c19